text/00110-compatibility-strategy.md

```diff
@@ -0,0 +1,364 @@
+We want to find a generic mechanism for running these tests, without having to cherry-pick use-cases or versions. 
+
+The main design idea is to introduce new types of integration tests into our pre-release suite. Those tests will run various permutations of `(CLI X Framework)`, and validate the expected result in each permutation.
+
+To make it generic, the tests will run the entire current integration tests suite, and simply inject different versions each time.
+
+An important note is that when testing compatibility to older versions (be it CLI or Framework), we cannot run the *latest* integration test suite, as it may contain tests for new features and bug fixes, which were not implement in older versions.
+
+Therefore, the entire design relies on us having the ability to:
+
+- *Run **old** integration tests using a **new** CLI version and an **old** Framework version.*
+- *Run **old** integration tests using a **new** Framework version and an **old** CLI version.*
+- *Mark a Framework version as incompatible*
+
+    - The tests should support an ability to specify versions who are not compatible, since the expected output would be different.
+
+# Detailed Design
+
+As mentioned, everything is based on our ability to run integration tests with various version permutations. We introduce a new function, `run_integration_tests_perm` which magically does what we need, as well as some useful constants.
+
+The design is best explained by simply looking at the proposed code changes. Most of what you need to know is in the inline comments of the code.
+
+**[common.bash](https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/test/integ/cli/common.bash)**
+
+```bash
+function bump_minor() {
+
+    version=$1
+
+    # Accept a version and bump the minor version.
+}
+
+function decrease_minor() {
+
+    version=$1
+
+    # Accept a version and decrease the minor version.
+
+}
+
+# This is the latest published version.
+CURRENT_VERSION=$(node -e 'console.log(require("package.json").version')
+
+# This is the soon to be published version.
+PREVIOUS_VERSION="$(decrease_minor ${CURRENT_VERSION})"
+
+# This is the previously published version.
+NEXT_VERSION="$(bump_minor ${CURRENT_VERSION})"
+
+function run_integration_tests_perm() {
+
+    integration_tests_version=1
+    framework_version=$2
+    cli_version=$3
+    expected_error=$4
+
+    # Checkout the relevant integration tests
+    # Install framework version
+    # Install CLI version
+    # Inject versions to integration tests
+    # Run tests (should either pass or print the expected error)
+
+}
+
+```
+
+We now create a new file, which is responsible for validating the compatibility of new framework versions against old cli versions. This test will have the following assertions:
+
+- Compatible versions must pass all integration tests.
+- Incompatible versions must print the expected error message.
+
```

Comment:
I have a feeling that maybe this is over-designing it a little. Wouldn't it be sufficient to simply run new CLI versions against the latest published framework version (and tests)? I have a feeling it should cover most of what we are after.

Reply:
That depends what compatibility assurance we are aiming for. What you suggested only verifies that new CLI is compatible with older framework versions. But currently, we are attempting to also provide the reverse compatibility (as long as the protocol version didn't change). To do that, we also have to run old CLI with new framework, and adjust the assertion according to the expected outcome.

Reply:
Not relevant anymore